File: LeetCode/Python/solutions/p0516.py
class Solution:
    def longestPalindromeSubseq(self, s: str) -> int:
        #
        # Solution v1: dynamic programming and backward
        #
        # Runtime: 1013 ms @ (beats) 87.43%
        # Memory Usage: 13.8 MB @ (beats) 99.94%
        #
        l = len(s)
        dp = [0] * l

        for i in range(l - 1, -1, -1):
            curr = [0] * l
            curr[i] = 1
            for j in range(i + 1, l):
                if s[i] == s[j]:
                    curr[j] = 2 + dp[j - 1]
                else:
                    curr[j] = max(dp[j], curr[j - 1])
            dp = curr

        return dp[-1]


if __name__ == "__main__":
    solution = Solution()

    s = "bbbab"
    print(f"return: {solution.longestPalindromeSubseq(s)}")  # 4

    s = "cbbd"
    print(f"return: {solution.longestPalindromeSubseq(s)}")  # 2

File: LeetCode/Java/src/solutions/p0516_LongestPalindromicSubsequence/P0516.java
package solutions.p0516_LongestPalindromicSubsequence;

public class P0516 {
    public int longestPalindromeSubseq(String s) {
        /*
         * Solution v1: Dynamic Programming
         *
         * Runtime: 39 ms @ (beats) 83.77%
         * Memory Usage: 42.7 MB @ (beats) 95.11%
         */
        int l = s.length();
        int[] dp = new int[l];
        for (int i = l - 1; i >= 0; i--) {
            int[] curr = new int[l];
            curr[i] = 1;
            for (int j = i + 1; j < l; j++) {
                if (s.charAt(i) == s.charAt(j)) {
                    curr[j] = 2 + dp[j - 1];
                } else {
                    curr[j] = Math.max(dp[j], curr[j - 1]);
                }
            }
            dp = curr;
        }
        return dp[l - 1];

    }
}
